### DISCUSSION TITLE
IPAdapter style transfer

### DISCUSSION THREAD

**Discussion by cubiq** [huggingface/diffusers#7534] [Ideas]
In SDXL by applying the weights only to the transformer index 6 it is possible to get a very powerful style transfer tool guided by IPAdapter. I don't know why I haven't thought about it before... I implemented it in [ComfyUI](https://github.com/cubiq/ComfyUI_IPAdapter_plus/) and I guess it would be a cool feature for diffusers.

Similarly I believe it's possible to do the same with the composition only (haven't worked on this yet)

In the corner you can see the reference image

"cyberpunk sports car"
![style_transfer_4](https://github.com/huggingface/diffusers/assets/427614/ee4b2f74-3660-49a3-81a7-ec9bdd1c6ca6)

"peaceful landscape in japan"
![style_transfer_3](https://github.com/huggingface/diffusers/assets/427614/78c9f485-7164-4c43-96e8-d15e550287ba)

"woman daydreaming at the window"
![style_transfer_2](https://github.com/huggingface/diffusers/assets/427614/b13d264f-1603-46b4-96b9-297b5fcba364)

"cyberpunk alley"
![style_transfer_1](https://github.com/huggingface/diffusers/assets/427614/9ae38028-9775-45c9-bf88-ff8a934b813c)

**Comment by asomoza** [comment]
Really interesting, I did some test in diffusers and it works really good and sometimes not that much, still need to test it more when I have more time.

I found that if I send as a negative the same image (with a little noise) it uses only the colors which is also nice.

Because I'm lazy, I used your car image as an input:

"a cat"
|normal|with negative image|
|---|---|
|![20240331205355](https://github.com/huggingface/diffusers/assets/5442875/974543b6-af93-4adf-8817-efabe5ce3a16)|![20240331205419](https://github.com/huggingface/diffusers/assets/5442875/022d0c79-fa7a-4923-a84f-f615dc425109)|

For example for this image taken from `Visual Style Prompting`

prompt: " a tiger"

|source|IP Adapter|Visual Style Prompting
|---|---|---|
|![ref_chinese-ink-paint_A horse](https://github.com/huggingface/diffusers/assets/5442875/7d09346e-24a6-4f22-9869-4e24820e38a1)|![20240331210516](https://github.com/huggingface/diffusers/assets/5442875/7c5a8bc4-c838-480f-9953-6d4c354da2b8)|![image](https://github.com/huggingface/diffusers/assets/5442875/e186c15e-d0d5-4127-ab66-10e3a22a9d6f)|

It doesn't capture the style as good but probably I can play with something to get to that point.

Thank you for this @cubiq , I will use it and it's really cool.

As a side note, My knowledge of the unet is not that deep yet, why is that you say:

> I don't know why I haven't thought about it before

Is it that obvious that the `transformers index 6` is the one responsible of the style? In that case which one(s) would be responsible for the composition?

Now I probably want a granular control over all of them if that's the case, the same as we have with LoRAs now.

**Reply by cubiq** [reply]
the prompt is very important, if you add just a few tokens you can get as close or as far as you want from the reference

![ComfyUI_temp_fdjii_00042_](https://github.com/huggingface/diffusers/assets/427614/31c73ece-a54b-4ed2-9f2e-602fbe428424)

**Reply by asomoza** [reply]
oh ok, that's one difference between the `visual style prompting`, still is close enough and at least in diffusers we don't  need to load another pipeline.

Also:

> Now I probably want a granular control over all of them if that's the case, the same as we have with LoRAs now.

I couldn't resist so I just did it, so now I know how to control the composition:

prompt: "a dog"

|source|generated|
|---|---|
|![20240331205355](https://github.com/huggingface/diffusers/assets/5442875/fe9ac7af-5de3-41db-ba61-30da3dc4b464)|![20240401050143](https://github.com/huggingface/diffusers/assets/5442875/2c2c4afe-54da-45b0-98f6-1459699aac62)|
|![20240331210516](https://github.com/huggingface/diffusers/assets/5442875/bf6bef8f-e4c9-4ce0-b939-24ec2f02c1f8)|![20240401050632](https://github.com/huggingface/diffusers/assets/5442875/f3ca76b5-112c-4b39-8323-7c52accda4c6)|

or how to make the dog more tigerish:

|without style|with style|
|---|---|
|![20240401051006](https://github.com/huggingface/diffusers/assets/5442875/48436b6f-b946-45d5-80a5-d6f6ad0aac66)|![20240401051131](https://github.com/huggingface/diffusers/assets/5442875/0b25c1c5-065d-4657-94cc-383f56f825ef)|

Really cool that we can make all this with just one ip adapter

**Comment by yiyixuxu** [comment]
super cool! can we add this as  a [community script](https://github.com/huggingface/diffusers/blob/main/examples/community/README_community_scripts.md) for now? we can figure out if we can better support all these experimental IP-adpter features later when we have more of the examples 

you can make a custom attention processor like this, basically just copy-paste `IPAdapterAttnProcessor2_0` and make the modification needed 

```python
class StyleIPAdapterAttnProcessor2_0(torch.nn.Module):
    def __init__(self):
        ...
    def forward(self, ...):
        ...
```

and then you can do something like this to set the processor 

```python
from diffusers import DiffusionPipeline
from diffusers.models.attention_processor import IPAdapterAttnProcessor, IPAdapterAttnProcessor2_0
import torch

pipe = DiffusionPipeline.from_pretrained(...)
pipe.load_ip_adapter(...)

your_attn_processor = pipe.unet.attn_processors
for k, w in your_attn_processor.items():
    if isinstance(w, (IPAdapterAttnProcessor, IPAdapterAttnProcessor2_0)):
        your_attn_processor[k] = StyleIPAdapterAttnProcessor2_0()

pipe.unet.set_attn_processor(your_attn_processor)
```

**Reply by cubiq** [reply]
I've done the same for composition instead of style and you can mix the two. It's pretty fun

![composition_3](https://github.com/huggingface/diffusers/assets/427614/1ca88c14-2803-444e-82d8-727a77c2431c)
![composition_2](https://github.com/huggingface/diffusers/assets/427614/915a0c14-54ee-4a0b-a95b-685b73fa803b)
![composition_1](https://github.com/huggingface/diffusers/assets/427614/755992c0-e7f7-4f9c-9fe2-f79d0803a752)

**Comment by darshats** [comment]
About this question - "Is it that obvious that the transformers index 6 is the one responsible of the style? In that case which one(s) would be responsible for the composition?"

I think [this ](https://b-lora.github.io/B-LoRA/)paper might explain? its the fifth transformer block index skipping the first block. 

@yardenfren1996

**Comment by Podidiving** [comment]
👋 

@asomoza can you share the code please? 🙏   
I tried to reproduce it and disabled `IP-adapter` for all  that are not `*up_blocks.0.attentions.1*` blocks but it didn't work as I was expecting

**Reply by asomoza** [reply]
Hi, this is old though, I did my own implementation for those generations. After that, what was added was InstantStyle which it should be the same. What result are you expecting?

**Reply by Podidiving** [reply]
> I did my own implementation for those generations

Could you share this implementation? I'm interested in the implementation you did for this discussion. I did mine, but I'm not sure it's the correct one, since I can't achieve quite the same results

**Reply by asomoza** [reply]
That implementation was a really quick hack I did back then, I don't have that code anymore. But it's basicaly what I have in the [app](https://github.com/ZCode-opensource/image-artisan-xl) I was doing back then:

![image](https://github.com/user-attachments/assets/2e16bc7a-4120-4359-891d-64e25a37093d)

![image](https://github.com/user-attachments/assets/cd26a343-0908-4100-8b94-71fda87af6c9)

The first one is without noise, the second one has a little gaussian noise. You can look at the code but that app uses a very heavy modification of diffusers with a node system and on top PyQT, so don't know if it's going to help you. [Here's](https://github.com/ZCode-opensource/image-artisan-xl/pull/38/files) the commit with that code.

**Reply by Podidiving** [reply]
ty ❤️
